segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryDeltaIterator.java

```diff
@@ -0,0 +1,188 @@
+import java.util.Map;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import lombok.Builder;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * Iterates through {@link TableBucket}s in a Segment.
+ * @param <T> Type of the final, converted result.
+ */
+@Slf4j
+@ThreadSafe
+@Builder
+class TableEntryDeltaIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    // The maximum size (in bytes) of each read to perform on the segment.
+    private static final int MAX_READ_SIZE = 2 * 1024 * 1024;
+
+    private final DirectSegmentAccess segment;
+    private final long startOffset;
+    private final int maxLength;
+    private final boolean shouldClear;
+    private final Duration fetchTimeout;
+    private final EntrySerializer entrySerializer;
+    private final ConvertResult<T> resultConverter;
+    private final Executor executor;
+
+    @GuardedBy("this")
+    private Iterator<Map.Entry<DeltaIteratorState, TableEntry>> currentEntry;
+    @GuardedBy("this")
+    private long currentBatchOffset;
+
+    //endregion
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        return getNextEntry()
+                .thenCompose(entry -> {
+                    if (entry == null) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return this.resultConverter.apply(entry);
+                    }
+                });
+    }
+
+    public synchronized boolean endOfSegment() {
+        return this.currentBatchOffset >= (this.startOffset + this.maxLength);
+    }
+
+    private CompletableFuture<Map.Entry<DeltaIteratorState, TableEntry>> getNextEntry() {
+        val entry = getNextEntryFromBatch();
+        if (entry != null) {
+            return CompletableFuture.completedFuture(entry);
+        }
+
+        return fetchNextTableEntriesBatch().thenApply(val -> getNextEntryFromBatch());
+    }
+
+    private synchronized Map.Entry<DeltaIteratorState, TableEntry> getNextEntryFromBatch() {
+        if (this.currentEntry != null) {
+            val next = this.currentEntry.next();
+            if (!this.currentEntry.hasNext()) {
+                this.currentEntry = null;
+            }
+            return next;
+        }
+
+        return null;
+    }
+
+    private CompletableFuture<Void> fetchNextTableEntriesBatch() {
+        return toEntries(currentBatchOffset)
+                .thenAccept(entries -> {
+                    if (!entries.isEmpty()) {
+                        this.currentEntry = entries.iterator();
```

Comment:
You are assigning `currentEntry` outside of declared guards. This is not thread safe.

Doesn't Intellij highlight this as a warning? It should be enabled by default.

Reply:
Nope, no warnings. Will look to enable it.